boundary.go

```diff
@@ -37,48 +43,127 @@ func newBoundaryReader(reader *bufio.Reader, boundary string) *boundaryReader {
+			return 0, errors.WithStack(err)
 		}
-	} else {
-		// No boundary found, move forward a safe distance
-		if nCopy = len(peek) - len(b.nlPrefix) - 1; nCopy <= 0 {
-			nCopy = 0
-			if peekEOF {
-				// No more peek space remaining and no boundary found
-				return 0, errors.WithStack(io.ErrUnexpectedEOF)
+		// Ensure that we can switch on the first byte of 'c' without panic
+		if len(c) > 0 {
+			switch c[0] {
+			// Check for line feed as potential LF boundary prefix
+			case '\n':
+				peek, err := b.r.Peek(len(b.nlPrefix) + 2)
+				switch err {
+				case nil:
+					// Check the whitespace at the head of the peek to avoid checking for a boundary early.
+					if bytes.HasPrefix(peek, []byte("\n\n")) ||
+						bytes.HasPrefix(peek, []byte("\n\r")) {
+						break
+					}
+					// Check the peek buffer for a boundary delimiter or terminator.
+					if b.isDelimiter(peek[1:]) || b.isTerminator(peek[1:]) {
+						// Check if we stored a carriage return.
+						if b.crBoundaryPrefix {
+							b.crBoundaryPrefix = false
+							// Let us now unread that back onto the io.Reader, since
+							// we have found what we are looking for and this byte
+							// belongs to the bounded block we are reading.
+							err = b.r.UnreadByte()
+							switch err {
+							case nil:
+								// Carry on
+							case bufio.ErrInvalidUnreadByte:
+								// Carriage return boundary prefix bit already unread
+							default:
+								return 0, errors.WithStack(err)
+							}
+						}
+						// We have found our boundary terminator, lets write out the final bytes
+						// and return io.EOF to indicate that this section read is complete.
+						n, err = b.buffer.Read(dest)
+						switch err {
+						case nil, io.EOF:
+							return n, io.EOF
+						default:
+							return 0, errors.WithStack(err)
+						}
+					}
+				case io.EOF:
+					// We have reached the end without finding a boundary,
+					// so we flag the boundary reader to add an error to
+					// the errors slice and write what we have to the buffer.
+					b.unbounded = true
+				default:
+					continue
+				}
+				// Checked '\n' was not prefix to a boundary.
+				if b.crBoundaryPrefix {
+					b.crBoundaryPrefix = false
+					// Stored '\r' should be written to the buffer now.
+					err = b.buffer.WriteByte('\r')
+					if err != nil {
+						return 0, errors.WithStack(err)
+					}
+				}
+			// Check for carriage return as potential CRLF boundary prefix
+			case '\r':
+				_, err := b.r.ReadByte()
+				if err != nil {
+					return 0, errors.WithStack(err)
+				}
+				// Flag the boundary reader to indicate that we
+				// have stored a '\r' as a potential CRLF prefix
+				b.crBoundaryPrefix = true
+				continue
 			}
 		}
-	}
-	if nCopy > 0 {
-		if _, err = io.CopyN(b.buffer, b.r, int64(nCopy)); err != nil {
-			return 0, errors.WithStack(err)
+
+		_, err = io.CopyN(b.buffer, b.r, 1)
+		if err != nil {
+			// EOF is not fatal, it just means
+			// that we have drained the reader.
```

Comment:
I think this comment can fit on one line.

Reply:
Done